### PR TITLE
refactor(BV): Merge `Congruence` and `Constraints`

### PR DESCRIPTION
**Pull request by bclement-ocp**
In https://github.com/OCamlPro/alt-ergo/pull/944, the `Congruence` module was added to simplify handling dependences for both domains and constraints, because we tracked domains separately for all terms.

After https://github.com/OCamlPro/alt-ergo/pull/1004, we now track domains only for uninterpreted leaves, and the `Congruence` module is only used for `Constraints`. This leads to a sort of double indirection: the `Congruence` module keeps track of reverse uninterpreted leaves -> class representative dependencies, and then the `Constraints` module keeps track of reverse class representative -> constraint dependencies.

This patch removes the `Congruence` module entirely; instead, the `Constraints` module is now keeping track of reverse uninterpreted leaves -> constraint dependencies directly.

In addition, the `Constraints` module is made into a functor and moved to the `Rel_utils` module, since it is now a generic module that can be useful for any relational theory (like `Congruence` was before).

I recommend reviewing the PR commit by commit to better be able to follow the changes: the commit that does the actual changes in implementation of the `Constraints` module is the one titled "Fold `Congruence` into `Constraints`"; the others are mostly shuffling code around, which makes the global diffs hard to follow.

**Note**: This currently includes #1004.

### CODE REVIEW

**Comment by bclement-ocp** [comment]
This is +1-0 on smtlib/BV; not sure it is worth running on QF_BV for this change since that takes a lot of time (maybe we should build a small subset of QF_BV).

Edit: +1-0 compared to #1004

File: src/lib/reasoners/rel_utils.ml
Comment (Halbaroth) [inline src/lib/reasoners/rel_utils.ml]:
```suggestion
  (** [subst ex p v cs] replaces all the instances of [p] with [v] in the
```

File: src/lib/reasoners/rel_utils.ml
Comment (Halbaroth) [inline src/lib/reasoners/rel_utils.ml]: I guess we expect to use this module for constraints in other theories? For instance, we should use it in ADT theory.

File: src/lib/reasoners/rel_utils.ml
Comment (Halbaroth) [inline src/lib/reasoners/rel_utils.ml]: Do you mean two identical representations with different explanations **have** to be equal for this comparison function?

File: src/lib/reasoners/rel_utils.ml
Comment (Halbaroth) [inline src/lib/reasoners/rel_utils.ml:359]: We use the polymorphic comparison here. It means the explanation is updated by this function but the `Constraint.compare` is supposed to ignore the explanation? Is it intentional?

We should at least mention in the documentation of `add` that the structure is updated if the explanation changed.

File: src/lib/reasoners/rel_utils.ml
Comment (Halbaroth) [inline src/lib/reasoners/rel_utils.ml:301]: Nit: why did you use three asterisks?

File: src/lib/reasoners/rel_utils.ml
Comment (Halbaroth) [inline src/lib/reasoners/rel_utils.ml]:
```suggestion
      new domain. *)
```

File: src/lib/reasoners/rel_utils.ml
Comment (Halbaroth) [inline src/lib/reasoners/rel_utils.ml]: Same here.

File: src/lib/reasoners/bitv_rel.ml
Comment (Halbaroth) [inline src/lib/reasoners/bitv_rel.ml]: It seems this PR is rebased on #1004, which means we cannot merge it before solving #1005 too.

File: src/lib/reasoners/rel_utils.ml
Comment (Halbaroth) [inline src/lib/reasoners/rel_utils.ml]: We should mention that the fresh constraint queue is flushed by this iterator.

File: src/lib/reasoners/rel_utils.ml
Comment (Halbaroth) [inline src/lib/reasoners/rel_utils.ml:305]: I don't understand the meaning of this field. We add a constraint to it in `add` and we flush it in `fold_fresh` but we don't touch it in `propagate`.

**Review by Halbaroth — CHANGES_REQUESTED** [review]

File: src/lib/reasoners/bitv_rel.ml
Comment (bclement-ocp) [reply src/lib/reasoners/bitv_rel.ml]: It is explicitly mentioned in the PR description that it includes #1004 :)

File: src/lib/reasoners/rel_utils.ml
Comment (bclement-ocp) [reply src/lib/reasoners/rel_utils.ml]: Yeah, I intended to remove reference to bit-vectors in this module but forgot some :)

File: src/lib/reasoners/rel_utils.ml
Comment (bclement-ocp) [reply src/lib/reasoners/rel_utils.ml]: They don't strictly have to, but they typically should. Actually I am not sure if it is very useful to bundle the explanations into the constraints, we might be able to have a single implementation for explanations (over a non-explained domain type). I'd rather do that in a separate PR to keep this one simple; in the meantime I will reformulate; this was written this way prior to being converted to a functor.

File: src/lib/reasoners/rel_utils.ml
Comment (bclement-ocp) [reply src/lib/reasoners/rel_utils.ml:359]: Where do we use polymorphic comparison? `CS` is defined as `Set.Make(Constraint)`, `CS.mem` uses `Constraint.compare`.

File: src/lib/reasoners/rel_utils.ml
Comment (bclement-ocp) [reply src/lib/reasoners/rel_utils.ml:305]: Because `propagate` propagates the constraints associated with a single representative `r`. The `fresh` field is cleared by `fold_fresh` which is intended to be used as `Constraints.fold_fresh Constraint.propagate` (see usage in `bitv_rel.ml`). I will refactor to expose `propagate_fresh` instead of `fold_fresh` which makes more sense.

File: src/lib/reasoners/rel_utils.ml
Comment (bclement-ocp) [reply src/lib/reasoners/rel_utils.ml:301]: It is a typo that was already present in #944, I will fix it

File: src/lib/reasoners/rel_utils.ml
Comment (bclement-ocp) [reply src/lib/reasoners/rel_utils.ml]: I will rename this function as `propagate_fresh` and only expose `fold_fresh Constraint.propagate` and mention this in the doc of that function.

File: src/lib/reasoners/rel_utils.ml
Comment (bclement-ocp) [reply src/lib/reasoners/rel_utils.ml]: The `domain` type is expected to contain domains for many variables; I will rename it (and add a short documentation as well).

**Comment by bclement-ocp** [comment]
Rebased after merging #1004; should be ready for a 2nd round @Halbaroth

**Review by Halbaroth — APPROVED** [review]

**Pull request merged**
